### DISCUSSION TITLE
Map AD to user..

### DISCUSSION THREAD

**Discussion by jakobvj** [Daniel-Krzyczkowski/Cars-Island-On-Azure#15] [Q&A]
Hi
What if i need much more information about the user. So when a user log in, i can get the user-object from a local database by the object_id from AD.
What is best practices for this?
Is it just to make a method GetUserById(objectId) after the user logs in?

**Comment by Daniel-Krzyczkowski** [comment]
Hi Jakob,

Let me answer your question - and if I misunderstood, please provide more details and I am happy to help and answer again.
Recommended way is to store complex user profile outside of Azure AD B2C. In this case, as you said, user data should be identified by ObjectID generated by Azure AD B2C when user account is created.
This value is unique so you can use it to link user profile data in the database. Please also keep in mind that this is recommended approach. Sometimes people link user profile in the database using user e-mail from Azure AD B2C but please keep in mind that once e-mail is changes (yes, email change is possible in the Azure AD B2C too), profile data in the database is no longer linked with user identity in the Azure AD B2C.
This is why recommended approach is to use ObjectID property assigned by Azure AD B2C to user account so you can easily create relation between user data in data stores and user identity in the Azure AD B2C.

I hope this clarifies a bit.

**Reply by jakobvj** [reply]
Thanks for the answer. Just what i was looking for. 

So, every time a user log in or want to do something, you take the objectId from the token and get the user from the database?

**Reply by justoke** [reply]
Hi Jakob,

I've recently built and deployed a live system using the B2C platform based on the work Daniel provided in this excellent resource that is Car Islands. It works really well. The main point here is to decouple your B2C from your application and specifically the customer in your application from the B2C user. By way of a simple data schema I can show you how I structured the data model. I like to think of the customer profile as the object that contains the object id that you know the user as from B2C and the customer id that your application cares about.  So when our customer logs into the application, we just take the object id and lookup their customer profile record and use that for everything the application requires. This simplifies the domain significantly.

{
    "userId": "0e89f5ad-de00-48f3-b8ac-cef516dd59cb",
    "customerName": "John",
    "phoneNumber": "+12345",
    "customerEmail": "any@somedomain.net",
    "status": "new,payment,completed",
    "id": "81adb3e6-79be-4937-9a02-3f6a15e82155",
}

Where userId is the B2C object id and id is the database identifier. We don't try to sync these two entities. In my use case, we use phone number for the sign up. 

I would not bother with keeping these two entities in sync.

**Reply by Daniel-Krzyczkowski** [reply]
Hi @justoke, thank you for this example here, very good one.
The case is that every time you need any data related to user from the database, you should use ObjectID for this correlation. When doing query on the database, you provide ObjectID (which should be extracted from the access token on the backend side) and you can be sure that you will get data only for specific user.
